packages/admin/admin/src/submissionBoundary/SubmissionBoundary.tsx

```diff
@@ -0,0 +1,137 @@
+    const [saving, setSaving] = React.useState(false);
+    const [hasErrors, setHasErrors] = React.useState(false);
+    const [hasChanges, setHasChanges] = React.useState(false);
+    const saveStates = React.useRef<Record<string, SubmissionBoundaryStateProps>>({});
+    const intl = useIntl();
+
+    const save = React.useCallback(async (): Promise<SaveActionSuccess> => {
+        setHasErrors(false);
+        setSaving(true);
+        try {
+            const saveSuccess = !(
+                await Promise.all(
+                    Object.values(saveStates.current).map((state) => {
+                        return state.doSave();
+                    }),
+                )
+            ).some((saveSuccess) => !saveSuccess);
+            if (!saveSuccess) {
+                setHasErrors(true);
+            } else {
+                onAfterSave?.();
+            }
+            return saveSuccess;
+        } catch (error: unknown) {
+            setHasErrors(true);
+            throw error;
+        } finally {
+            setSaving(false);
+        }
+    }, [onAfterSave]);
+
+    const onSaveStatesChanged = React.useCallback(() => {
+        const hasChanges = Object.values(saveStates.current).some((saveState) => saveState.hasChanges);
+        setHasChanges(hasChanges);
+    }, []);
+
+    const register = React.useCallback(
+        (id: string, props: SubmissionBoundaryStateProps) => {
+            saveStates.current[id] = props;
+            onSaveStatesChanged();
+        },
+        [onSaveStatesChanged],
+    );
+    const unregister = React.useCallback(
+        (id: string) => {
+            delete saveStates.current[id];
+            onSaveStatesChanged();
+        },
+        [onSaveStatesChanged],
+    );
+
+    return (
+        <RouterPrompt
+            message={() => {
+                if (hasChanges) {
+                    return intl.formatMessage(messages.saveUnsavedChanges);
+                }
+                return true;
+            }}
+            saveAction={save}
+            subRoutePath={props.subRoutePath}
+        >
+            <SubmissionBoundaryStateContext.Provider
+                value={{
+                    hasErrors,
+                    hasChanges,
+                    saving,
+                }}
+            >
+                <SubmissionBoundaryApiContext.Provider
+                    value={{
+                        save,
+                        register,
+                        unregister,
+                    }}
+                >
+                    {props.children}
+                </SubmissionBoundaryApiContext.Provider>
+            </SubmissionBoundaryStateContext.Provider>
+        </RouterPrompt>
+    );
+}
+
+export interface SubmissionBoundaryStateProps {
+    hasChanges: boolean;
+    doSave: () => Promise<SaveActionSuccess> | SaveActionSuccess;
+}
+
+export function SubmissionBoundaryState({ doSave, hasChanges }: SubmissionBoundaryStateProps) {
```

Comment:
I think a better name would be `RegisterToSubmissionBoundary`, what do you think? It's not really a state...

Reply:
yes, but lets discuss the naming here https://github.com/vivid-planet/comet/pull/1448/files/e11351df4a90984f61a92acf2369985edd0458e0#r1430094463